### PR TITLE
Reminders

### PR DESCRIPTION
**Pull request by yuhongtay**
Sorry

### CODE REVIEW

File: src/main/java/seedu/guilttrip/logic/commands/remindercommands/SetReminderCommand.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+/**
+ * Set a reminder for an expense/ income/ wish.
+ */
+public class SetReminderCommand extends Command {
+    public static final String COMMAND_WORD = "setReminder";
+    public static final String UNSUPPORTED_TYPE = "Unable to create reminder for this entry.";
+    public static final String INVALID_PERIOD = "Invalid period. (Cannot set reminder to notify before today!)";
+    public static final String INVALID_FREQ = "Frequency must be smaller than period.";
+    public static final String MESSAGE_SUCCESS = "New reminder added for: %1$s";
+    public static final String MESSAGE_USAGE = COMMAND_WORD
```
Comment (SwampertX) [inline src/main/java/seedu/guilttrip/logic/commands/remindercommands/SetReminderCommand.java]: Can you refactor the one-liner description into a variable? And add it into the map function in `ParserUtil.java` if you want your command to be reflected in the AutoSuggester.

You can check out any other command for an example!

File: src/main/java/seedu/guilttrip/logic/commands/remindercommands/TogglePopUpCommand.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import seedu.guilttrip.logic.commands.CommandResult;
+import seedu.guilttrip.logic.commands.exceptions.CommandException;
+import seedu.guilttrip.model.Model;
+
+/**
+ * Toggles pop up message of generalReminder.
+ */
+public class TogglePopUpCommand extends Command {
+
+    public static final String COMMAND_WORD = "showPopUp";
```
Comment (SwampertX) [inline src/main/java/seedu/guilttrip/logic/commands/remindercommands/TogglePopUpCommand.java]: Here also.

File: src/main/java/seedu/guilttrip/logic/commands/remindercommands/SelectReminderCommand.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import seedu.guilttrip.model.reminders.Reminder;
+
+/**
+ * Select the generalReminder for editing.
+ */
+public class SelectReminderCommand extends Command {
+    public static final String COMMAND_WORD = "selectReminder";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ": Select the reminder identified by the index number used in the "
```
Comment (SwampertX) [inline src/main/java/seedu/guilttrip/logic/commands/remindercommands/SelectReminderCommand.java]: Here also.

File: src/main/java/seedu/guilttrip/logic/commands/remindercommands/DisplayPopupCommand.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import seedu.guilttrip.logic.commands.Command;
+import seedu.guilttrip.logic.commands.CommandResult;
+import seedu.guilttrip.logic.commands.exceptions.CommandException;
+import seedu.guilttrip.model.Model;
+
+/**
+ * Displays popup for GeneralReminder.
+ */
+public class DisplayPopupCommand extends Command {
+    public static final String COMMAND_WORD = "popup";
```
Comment (SwampertX) [inline src/main/java/seedu/guilttrip/logic/commands/remindercommands/DisplayPopupCommand.java]: Same here. You already have a great description at the javadoc for this class, can just copy.

File: src/main/java/seedu/guilttrip/logic/commands/remindercommands/AutoDisplayPopUp.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import seedu.guilttrip.logic.commands.Command;
+import seedu.guilttrip.logic.commands.CommandResult;
+import seedu.guilttrip.logic.commands.exceptions.CommandException;
+import seedu.guilttrip.model.Model;
+import seedu.guilttrip.model.reminders.messages.Message;
+
+/**
+ * automatically displays message popup.
+ */
+public class AutoDisplayPopUp extends Command {
```
Comment (SwampertX) [inline src/main/java/seedu/guilttrip/logic/commands/remindercommands/AutoDisplayPopUp.java]: Do you want to rename this file to `AutoDisplayPopUpCommand.java`?

File: src/main/java/seedu/guilttrip/logic/commands/remindercommands/AddGeneralReminderCommand.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+
+/**
+ * Adds a entry to the guilttrip book.
+ */
+public class AddGeneralReminderCommand extends Command {
+
+    public static final String COMMAND_WORD = "addReminder";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Adds a GeneralReminder to reminders list. "
```
Comment (SwampertX) [inline src/main/java/seedu/guilttrip/logic/commands/remindercommands/AddGeneralReminderCommand.java]: Same here. Tip: you can use <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>V</kbd> for abstracting the highlighted element into a variable.

File: src/main/java/seedu/guilttrip/logic/commands/SetCurrDateCommand.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+import seedu.guilttrip.logic.CommandHistory;
+import seedu.guilttrip.logic.commands.exceptions.CommandException;
+import seedu.guilttrip.model.Model;
+import seedu.guilttrip.model.entry.Date;
+
+/**
+ * FOrce sets the date for testing purposes.
+ */
+public class SetCurrDateCommand extends Command {
+    public static final String COMMAND_WORD = "setDate";
```
Comment (SwampertX) [inline src/main/java/seedu/guilttrip/logic/commands/SetCurrDateCommand.java]: Here. can check toggle commands as a reference.

File: src/main/java/seedu/guilttrip/logic/commands/remindercommands/RemoveConditionFromReminderCommand.java [hunk 8]
```diff
@@ -1,18 +1,19 @@
...
+import seedu.guilttrip.model.reminders.GeneralReminder;
 import seedu.guilttrip.model.reminders.Reminder;
 import seedu.guilttrip.model.reminders.conditions.Condition;
+import seedu.guilttrip.model.reminders.conditions.DateCondition;
+import seedu.guilttrip.model.reminders.conditions.QuotaCondition;
+import seedu.guilttrip.model.reminders.conditions.TagsCondition;
 
 /**
- * Removes condition from reminder.
+ * Removes condition from generalReminder.
```
Comment (SwampertX) [inline src/main/java/seedu/guilttrip/logic/commands/remindercommands/RemoveConditionFromReminderCommand.java]: Same here.

File: src/main/java/seedu/guilttrip/logic/commands/remindercommands/WriteMessageCommand.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+import seedu.guilttrip.model.Model;
+
+/**
+ * Writes message in generalReminder popup.
+ */
+public class WriteMessageCommand extends Command {
+    public static final String COMMAND_WORD = "writeMessage";
+    public static final String IMAGE_ADDED_MESSAGE = "Message successfully written to popup";
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ": Write message in popup of reminderExamined.\n"
```
Comment (SwampertX) [inline src/main/java/seedu/guilttrip/logic/commands/remindercommands/WriteMessageCommand.java]: here

File: src/main/java/seedu/guilttrip/logic/parser/remindercommandparsers/AddGeneralReminderCommandParser.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+     * Parses the given {@code String} of arguments in the context of the AddCommand
+     * and returns an AddCommand object for execution.
+     * @throws ParseException if the user input does not conform the expected format
+     */
+    public AddGeneralReminderCommand parse(String args) throws ParseException {
+        ArgumentMultimap argMultimap =
+                ArgumentTokenizer.tokenize(
+                        args, PREFIX_TYPE, PREFIX_DESC, PREFIX_UPPER_BOUND, PREFIX_LOWER_BOUND,
+                        PREFIX_START_DATE, PREFIX_END_DATE, PREFIX_TAG);
+        if (!arePrefixesPresent(argMultimap, PREFIX_DESC)) {
```
Comment (SwampertX) [inline src/main/java/seedu/guilttrip/logic/parser/remindercommandparsers/AddGeneralReminderCommandParser.java]: [Optional]: You can use the `throwError....()` function in `ParserUtil.java` for more expressive error messages. It's up to you here

File: src/main/java/seedu/guilttrip/logic/parser/remindercommandparsers/SetReminderCommandParser.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    public SetReminderCommand parse(String args) throws ParseException {
+        requireNonNull(args);
+        ArgumentMultimap argMultimap =
+                ArgumentTokenizer.tokenize(args, PREFIX_TYPE, PREFIX_DESC, PREFIX_PERIOD, PREFIX_FREQ);
+        try {
+            index = ParserUtil.parseIndex(argMultimap.getPreamble());
+        } catch (ParseException pe) {
+            throw new ParseException(String.format(MESSAGE_INVALID_ARGUMENT_FORMAT, SetReminderCommand.MESSAGE_USAGE));
+        }
+        if (!arePrefixesPresent(argMultimap, PREFIX_TYPE, PREFIX_DESC, PREFIX_PERIOD, PREFIX_FREQ)) {
```
Comment (SwampertX) [inline src/main/java/seedu/guilttrip/logic/parser/remindercommandparsers/SetReminderCommandParser.java]: You can use the helper function from parserutil too.

File: src/main/java/seedu/guilttrip/model/Model.java [hunk 12]
```diff
@@ -30,15 +32,15 @@
 public interface Model {
     /** {@code Predicate} that always evaluate to true */
     Predicate<Entry> PREDICATE_SHOW_ALL_ENTRIES = unused -> true;
+
```
Comment (SwampertX) [inline src/main/java/seedu/guilttrip/model/Model.java]: This empty line is weird though, can remove

File: src/main/java/seedu/guilttrip/model/entry/SortType.java [hunk 13]
```diff
@@ -25,7 +25,10 @@
     /**
      * Constructs a {@code SortType}.
      *
+<<<<<<< HEAD
+=======
      * @param fullType A valid sortType.
+>>>>>>> f659a201d941a9d7dedf6472b89947095b946820
```
Comment (SwampertX) [inline src/main/java/seedu/guilttrip/model/entry/SortType.java]: Please resolve this merge conflict.

File: src/main/java/seedu/guilttrip/model/reminders/GeneralReminder.java [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+    private long numberOfConditionsMet = 0;
+    private Status status = Status.unmet;
+
+    private boolean displayPopUp = false;
+    private Message message;
+
+    private final Logger logger = LogsCenter.getLogger(getClass());
+    private ObservableSupport support = new ObservableSupport();
+
+    //===== Constructor =====//
```
Comment (SwampertX) [inline src/main/java/seedu/guilttrip/model/reminders/GeneralReminder.java]: I love these comment headings!

File: src/test/java/seedu/guilttrip/logic/commands/FindExpenseCommandTest.java [hunk 15]
```diff
@@ -84,20 +85,6 @@ public void execute_multipleKeywords_multipleExpensesFound() {
...
-        EntryContainsDescriptionPredicate predicate = preparePredicate("mala cotton Mars");
-        EntryContainsAmountPredicate amtpredicate = new EntryContainsAmountPredicate(500);
-        List<Predicate<Entry>> predicateToPassIn = Arrays.asList(predicate, amtpredicate);
-        FindExpenseCommand command = new FindExpenseCommand(predicateToPassIn);
-        expectedModel.updateFilteredExpenses(predicate.and(amtpredicate));
-        assertCommandSuccess(command, model, expectedMessage, expectedModel, chs);
-        assertEquals(Arrays.asList(TypicalEntries.TRAVEL_EXPENSE),
-                model.getFilteredExpenses());
-    }*/
-
```
Comment (SwampertX) [inline src/test/java/seedu/guilttrip/logic/commands/FindExpenseCommandTest.java]: Any idea why this test failed?

File: src/test/java/seedu/guilttrip/testutil/TypicalEntries.java [hunk 16]
```diff
@@ -61,6 +61,10 @@
             .withCatName("travel").build();
     public static final Category CATEGORY_BUSINESS = new CategoryBuilder().withCatType("Income").withCatName("business")
             .build();
+
+    /*public static final Category CATEGORY_STOCKS = new CategoryBuilder().withCatType("Income").withCatName("stocks")
+            .build();*/
```
Comment (SwampertX) [inline src/test/java/seedu/guilttrip/testutil/TypicalEntries.java]: Seems to be similar to previous tests, any idea why this failed?

File: build.gradle [hunk 17]
```diff
@@ -59,6 +59,9 @@ dependencies {
     implementation group: 'org.openjfx', name: 'javafx-graphics', version: javaFxVersion, classifier: 'win'
     implementation group: 'org.openjfx', name: 'javafx-graphics', version: javaFxVersion, classifier: 'mac'
     implementation group: 'org.openjfx', name: 'javafx-graphics', version: javaFxVersion, classifier: 'linux'
+    implementation group: 'org.openjfx', name: 'javafx-swing', version: javaFxVersion, classifier: 'win'
+    implementation group: 'org.openjfx', name: 'javafx-swing', version: javaFxVersion, classifier: 'mac'
+    implementation group: 'org.openjfx', name: 'javafx-swing', version: javaFxVersion, classifier: 'linux'
```
Comment (SwampertX) [inline build.gradle]: Can I ask where is swing used? Can't seem to find it in the changes.

**Comment by SwampertX** [comment]
This PR is unfortunately too huge to review in any case. Our last resort (due to poor test coverage):
Please comment below if you have done the following:

Step 1: Clone the repo following the command line instructions
Step 2: Test your own feature - ensure things are working
Step 3: Make a comment below if you have tested it out and things work. otherwise request changes by tmr 10am

File: src/main/java/seedu/guilttrip/logic/commands/remindercommands/EditReminderCommand.java [hunk 18]
```diff
@@ -245,11 +336,48 @@ public boolean equals(Object other) {
...
+            return getLowerBound().equals(e.getLowerBound())
+                    && getUpperBound().equals(e.getUpperBound())
+                    && getStart().equals(e.getStart())
+                    && getEnd().equals(e.getEnd())
+                    && getTags().equals(e.getTags());
+        }
+    }
+
+    /**
+     * Contains edited information for IewReminder
```
Comment (Dandford) [inline src/main/java/seedu/guilttrip/logic/commands/remindercommands/EditReminderCommand.java]: Perhaps you would want to change the Iew to a longer more intuitive name? it's ok for the command to be longer as long as it is understandable

File: src/main/java/seedu/guilttrip/ui/ResultDisplay.java [hunk 19]
```diff
@@ -21,6 +24,8 @@ public ResultDisplay() {
     }
 
     public void setFeedbackToUser(String feedbackToUser) {
+        logger.info("Sending feedback: " + feedbackToUser);
+        logger.info("filling inner parts");
```
Comment (jessicax941) [inline src/main/java/seedu/guilttrip/ui/ResultDisplay.java]: Why is this logger logging `filling inner parts` here?

File: src/main/java/seedu/guilttrip/logic/commands/remindercommands/SetReminderCommand.java (hunk 1 above)
Comment (yuhongtay) [reply src/main/java/seedu/guilttrip/logic/commands/remindercommands/SetReminderCommand.java]: Actually if you check my Commands already have command word one liner and message usage.

File: build.gradle (hunk 17 above)
Comment (yuhongtay) [reply build.gradle]: Wanted to store images at one point

File: src/main/java/seedu/guilttrip/logic/commands/remindercommands/TogglePopUpCommand.java (hunk 2 above)
Comment (yuhongtay) [reply src/main/java/seedu/guilttrip/logic/commands/remindercommands/TogglePopUpCommand.java]: fix

File: src/main/java/seedu/guilttrip/model/reminders/GeneralReminder.java (hunk 14 above)
Comment (yuhongtay) [reply src/main/java/seedu/guilttrip/model/reminders/GeneralReminder.java]: fix

File: src/main/java/seedu/guilttrip/logic/commands/remindercommands/EditReminderCommand.java (hunk 18 above)
Comment (yuhongtay) [reply src/main/java/seedu/guilttrip/logic/commands/remindercommands/EditReminderCommand.java]: done

File: src/main/java/seedu/guilttrip/ui/ResultDisplay.java (hunk 19 above)
Comment (yuhongtay) [reply src/main/java/seedu/guilttrip/ui/ResultDisplay.java]: can delete

File: src/test/java/seedu/guilttrip/logic/commands/FindExpenseCommandTest.java (hunk 15 above)
Comment (yuhongtay) [reply src/test/java/seedu/guilttrip/logic/commands/FindExpenseCommandTest.java]: Works now.

Comment (SwampertX) [reply src/test/java/seedu/guilttrip/logic/commands/FindExpenseCommandTest.java]: Do you mind bringing this test back? thanks.

**Pull request merged**